src/python/pants/backend/shell/shunit2_test_runner.py

```diff
@@ -180,24 +180,16 @@ async def setup_shunit2_for_target(
     )
 
     field_set_digest_content = await Get(DigestContents, Digest, field_set_sources.snapshot.digest)
-    # Because a FieldSet corresponds to a file address, there should be exactly 1 file in the
-    # sources. This assumption allows us to simplify determining which shell to use via inspecting
-    # the shebang.
-    if len(field_set_digest_content) != 1:
-        raise AssertionError(
-            f"The file address {request.field_set.address} had sources != 1, which is unexpected: "
-            f"{field_set_sources.snapshot.files}. Please file a bug at "
-            "https://github.com/pantsbuild/pants/issues/new with this error message copied."
-        )
-    original_test_file_content = field_set_digest_content[0]
-    updated_test_file_content = add_source_shunit2(original_test_file_content)
+    # `ShellTestSourceField` validates that there's exactly one file.
+    test_file_content = field_set_digest_content[0]
+    updated_test_file_content = add_source_shunit2(test_file_content)
```

Comment:
I'm excited about this cleanup in particular. Rule authors don't need to know about "file targets" anymore, or even generated targets! The rule behaves the same whether the file-based target was generated or explicitly declared.